File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int dp(int n, vector<int> &memo) {
        if(n <= 2) {
            memo[n] = n;
            return n;
        }
        if(memo[n] > 0) {
            return memo[n];
        }
        memo[n] = dp(n-1, memo) + dp(n-2, memo);
        return memo[n];
    }
    
    int climbStairs(int n) {
        vector<int>memo(n+1, 0);
        return dp(n, memo);   
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.kt
class Solution {
    private val memo = Array<Int>(46) {0}
    fun dp(n: Int): Int {
        if(n <= 2) {
            return n
        }
        if( memo[n] > 0) {
            return memo[n]
        }
    
        memo[n] = dp(n-1) + dp(n-2)
        return memo[n]
    }
    fun climbStairs(n: Int): Int {
        return dp(n)
    }
}